File: admin-app/services/semaforo.py
"""Semáforo de voluntad de pago (escala fija de 5 colores)."""

from __future__ import annotations

# id → (label, color hex)
NIVELES: dict[str, tuple[str, str]] = {
    "rojo": ("Renuente", "#DC2626"),
    "naranja": ("Resistente", "#EA580C"),
    "amarillo": ("Indefinido", "#EAB308"),
    "lima": ("Dispuesto", "#84CC16"),
    "verde": ("Cooperativo", "#16A34A"),
}

IDS_ORDEN = ("rojo", "naranja", "amarillo", "lima", "verde")

# Valor especial para filtros (no se persiste en el cliente)
SIN_CLASIFICAR = "__sin__"


def normalize_semaforo(value: str | None) -> str:
    """Return a valid level id or empty string."""
    if not value:
        return ""
    key = str(value).strip().lower()
    return key if key in NIVELES else ""


def label_for(value: str | None) -> str:
    key = normalize_semaforo(value)
    if not key:
        return "Sin clasificar"
    return NIVELES[key][0]


def color_for(value: str | None) -> str:
    key = normalize_semaforo(value)
    if not key:
        return "#94A3B8"
    return NIVELES[key][1]


def filter_options() -> list[tuple[str, str]]:
    """(value, label) for dropdowns including Sin clasificar."""
    opts = [(SIN_CLASIFICAR, "Sin clasificar")]
    for sid in IDS_ORDEN:
        opts.append((sid, NIVELES[sid][0]))
    return opts

File: flutter-app/lib/models/semaforo.dart
import 'package:flutter/material.dart';

/// Escala fija de voluntad de pago (un valor por cliente).
class SemaforoNivel {
  final String id;
  final String label;
  final Color color;

  const SemaforoNivel({
    required this.id,
    required this.label,
    required this.color,
  });
}

/// Id especial solo para filtros (no se persiste).
const String kSemaforoSinClasificar = '__sin__';

const List<SemaforoNivel> kSemaforoNiveles = [
  SemaforoNivel(id: 'rojo', label: 'Renuente', color: Color(0xFFDC2626)),
  SemaforoNivel(id: 'naranja', label: 'Resistente', color: Color(0xFFEA580C)),
  SemaforoNivel(id: 'amarillo', label: 'Indefinido', color: Color(0xFFEAB308)),
  SemaforoNivel(id: 'lima', label: 'Dispuesto', color: Color(0xFF84CC16)),
  SemaforoNivel(id: 'verde', label: 'Cooperativo', color: Color(0xFF16A34A)),
];

SemaforoNivel? findSemaforo(String? id) {
  if (id == null || id.isEmpty) return null;
  for (final n in kSemaforoNiveles) {
    if (n.id == id) return n;
  }
  return null;
}

String normalizeSemaforo(String? raw) {
  final key = (raw ?? '').trim().toLowerCase();
  return findSemaforo(key)?.id ?? '';
}

String labelSemaforo(String? id) {
  final n = findSemaforo(id);
  return n?.label ?? 'Sin clasificar';
}

Color colorSemaforo(String? id) {
  final n = findSemaforo(id);
  return n?.color ?? const Color(0xFF94A3B8);
}
